custom_ops.py

```diff
@@ -0,0 +1,78 @@
+import torch
+import numpy as np
+# aka torch.library
+from library import OpDef, traceable, device_types
+
+# =====================================================================
+# This is example user library code. It defines
+# my_sin and my_sin_cos operators.
+# =====================================================================
+
+# User provides their custom op schema and implementations
+class MySin(OpDef):
+    schema = "(Tensor x) -> Tensor"
+
+    # the black-box cpu kernel
+    @staticmethod
+    def impl_cpu(x):
+        return torch.from_numpy(np.sin(x.detach().cpu().numpy()))
+
+    # the black-box cuda kernel
+    @staticmethod
+    def impl_cuda(x):
+        return torch.from_numpy(np.sin(x.detach().cpu().numpy())).to(x.device)
+
+    # the abstract impl. Must be "traceable". User must use opcheck to test.
+    @staticmethod
+    def abstract(x):
+        return torch.empty_like(x)
+
+    # autograd: provide us setup_backward() and backward() methods.
+    # these must be "traceable". User must use opcheck to test.
+    @staticmethod
+    def setup_backward(ctx, inputs, output):
+        x, = inputs
+        ctx.save_for_backward(x)
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        x, = ctx.saved_tensors 
+        return grad_output * x.cos()
+
+
+# Builds the provided implementations into a "custom op" that behaves like a function.
+# We don't do something like MySin.apply (i.e. the autograd.Function way) because
+# dynamic changes to methods on the MySin object will not be included in the op.
+my_sin_op = MySin.build()
+
+
+# The user should provide a function that calls the custom op so they can add a docstring.
+def my_sin(x):
+    """my_sin(x: Tensor) -> Tensor
+
+    Returns the sin of x.
+    """
+    return my_sin_op(x)
```

Comment:
Blehhhh boilerplate. Why not pass the docstring into build?

Reply:
It's just a couple more lines and it seems more natural this way. If we shove everything into build, then:
- does the user need to pass the type signature as well? (I guess we could generate it)
- I want to ban kwarg-only arguments in schemas for custom ops created with this API. Having the wrapper allows users to specify kwarg-only arguments.